### PR TITLE
refactor(logical agg): dedup columns in materialized input agg state …

### PR DESCRIPTION
**Pull request by soundOfDestiny**
…table

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

dedup columns in materialized input agg state table
previously `group_key` and upstream `pk_key` may overlap

implement a todo that @st1page wrote

## Checklist

- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

**Comment by st1page** [comment]
please run the `./risedev do-apply-planner-test` and we can see what changes on our plan.

**Comment by soundOfDestiny** [comment]
> please run the `./risedev do-apply-planner-test` and we can see what changes on our plan.

yes and I have just checked the new plans

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 1]
```diff
@@ -133,27 +133,31 @@ impl LogicalAgg {
                 TableCatalogBuilder::new(self.ctx().inner().with_options.internal_table_subset());
             let mut column_mapping = vec![];
 
+            let mut upstream_idx_to_state_idx = BTreeMap::new();
```
Comment (richardchien) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:136]: IIUC, `BTreeSet` or `HashSet` is enough?

Comment (soundOfDestiny) [reply src/frontend/src/optimizer/plan_node/logical_agg.rs]: > IIUC, `BTreeSet` or `HashSet` is enough?

sorry, I once thought that we need this column so I use `BTreeMap`, after that I recognized that we need not, but I forget to replace `BTreeMap` with `BTreeSet`. fixed.

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 2]
```diff
@@ -133,27 +133,31 @@ impl LogicalAgg {
                 TableCatalogBuilder::new(self.ctx().inner().with_options.internal_table_subset());
             let mut column_mapping = vec![];
 
+            let mut upstream_idx_to_state_idx = BTreeSet::new();
```
Comment (richardchien) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:136]: I think the variable name should be changed accordingly😇. Maybe even no need to `mut` and `insert`, just `let group_keys = BTreeSet::from_iter(self.group_key().iter().cloned())` is ok.

Comment (soundOfDestiny) [reply src/frontend/src/optimizer/plan_node/logical_agg.rs]: > I think the variable name should be changed accordingly😇. Maybe even no need to `mut` and `insert`, just `let group_keys = BTreeSet::from_iter(self.group_key().iter().cloned())` is ok.

fixed

File: src/frontend/src/optimizer/plan_node/logical_agg.rs [hunk 3]
```diff
@@ -133,6 +133,14 @@ impl LogicalAgg {
...
             let mut column_mapping = vec![];
 
+            let mut stored_upstream_idx = self
+                .group_key()
+                .iter()
+                .cloned()
+                .chain(sort_keys.iter().map(|&(_, idx)| idx))
+                .collect_vec();
+            stored_upstream_idx.sort();
+
```
Comment (richardchien) [inline src/frontend/src/optimizer/plan_node/logical_agg.rs:143]: What about:

```rust
            let mut included_upstream_indices = HashSet::new();
            let mut add_column = |upstream_idx, order_type| {
                if !included_upstream_indices.contains(&upstream_idx) {
                    let tb_column_idx =
                        internal_table_catalog_builder.add_column(&in_fields[upstream_idx]);
                    column_mapping.push(upstream_idx);
                    if let Some(order_type) = order_type {
                        internal_table_catalog_builder.add_order_column(tb_column_idx, order_type);
                    }
                    included_upstream_indices.insert(upstream_idx);
                }
            };

            for &idx in self.group_key() {
                add_column(idx, Some(OrderType::Ascending));
            }
            for (order_type, idx) in sort_keys {
                add_column(idx, Some(order_type));
            }
            for &idx in &in_pks {
                add_column(idx, Some(OrderType::Ascending));
            }
            for idx in include_keys {
                add_column(idx, None);
            }
```

May need to rerun `do-apply-planner-test`.

Comment (soundOfDestiny) [reply src/frontend/src/optimizer/plan_node/logical_agg.rs:143]: > What about:
> 
> ```rust
>             let mut included_upstream_indices = HashSet::new();
>             let mut add_column = |upstream_idx, order_type| {
>                 if !included_upstream_indices.contains(&upstream_idx) {
>                     let tb_column_idx =
>                         internal_table_catalog_builder.add_column(&in_fields[upstream_idx]);
>                     column_mapping.push(upstream_idx);
>                     if let Some(order_type) = order_type {
>                         internal_table_catalog_builder.add_order_column(tb_column_idx, order_type);
>                     }
>                     included_upstream_indices.insert(upstream_idx);
>                 }
>             };
> 
>             for &idx in self.group_key() {
>                 add_column(idx, Some(OrderType::Ascending));
>             }
>             for (order_type, idx) in sort_keys {
>                 add_column(idx, Some(order_type));
>             }
>             for &idx in &in_pks {
>                 add_column(idx, Some(OrderType::Ascending));
>             }
>             for idx in include_keys {
>                 add_column(idx, None);
>             }
> ```
> 
> , which will make the code cleaner.
> 
> May need to rerun `do-apply-planner-test`.

It's incorrect when we call `add_column(x, None)` before calling `add_column(x, Some(_))`

Comment (soundOfDestiny) [reply src/frontend/src/optimizer/plan_node/logical_agg.rs:143]: > What do you mean by calling `add_column(x, None)` before calling `add_column(x, Some(_))`.🤔 The logic is not changed compared to the previous code.

I mean if someone misuses this `add_column`, and he calls `add_column(x, None)` before calling `add_column(x, Some(_))`

Comment (richardchien) [reply src/frontend/src/optimizer/plan_node/logical_agg.rs:143]: But `add_column` is a temporary closure. It's not to hide the implementation, but just to simplify code. So devs should know what `add_column` does, just like before.

**Review by richardchien — APPROVED** [review]

**Pull request merged**